File: 49-group-anagrams/group-anagrams.java
class Solution {
    // Sorting Solution - TC: O(m*nlogn), SC: O(m)
    public List<List<String>> groupAnagrams(String[] strs) {
        // Map the sorted version of each word in strs (key) to its anagrams (value)
        Map<String, List<String>> anagramMap = new HashMap<>(); // SC: O(m)

        // TC: O(m), m is the number of strings in strs
        for (String str : strs) {
            char[] chars = str.toCharArray();
            Arrays.sort(chars); // TC: O(nlogn), n is the avg word length of each string
            String sortedStr = new String(chars);

            if (!anagramMap.containsKey(sortedStr)) {
                anagramMap.put(sortedStr, new ArrayList<>());
            }
            anagramMap.get(sortedStr).add(str);
        }

        return new ArrayList<>(anagramMap.values());
    }
}

File: 49-group-anagrams/group-anagrams.py
class Solution:
    # Optimal Hashmap Solution
    # TC: O(m*n*26) = O(m*n) where m = num of strs and n = avg len of each str
    # SC: O(26) = O(1)
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # map char count of each str to list of anagrams
        anagram_map = collections.defaultdict(list)

        for s in strs:
            count = [0] * 26

            for c in s:
                count[ord(c) - ord('a')] += 1
            
            # change count (key) to tuple because lists cannot be keys
            anagram_map[tuple(count)].append(s)
        
        return list(anagram_map.values())

    # Sorting + Hashmap Solution
    # TC: O(m*nlogn) where m = num of strs and n = avg len of each str
    # SC: O(n)
    def groupAnagrams2(self, strs: List[str]) -> List[List[str]]:
        anagram_map = collections.defaultdict(list)

        for s in strs:
            # use sorted version of each string as key for map
            sorted_str = ''.join(sorted(s))
            anagram_map[sorted_str].append(s)
        
        return list(anagram_map.values())
